# Import Libraries

In [34]:
# dataframe libraries
import pandas as pd
import numpy as np

# model load
import pickle

# Imput Data Inference

Menambahkan data inference sebagai data uji, dimana data yg diuji berisi data yang belum pernah dilihat oleh model

In [35]:

# Original data dictionary
data = {
    'HomeTeam': ['Charlton', 'Chelsea', 'Coventry', 'Derby', 'Leeds'],
    'AwayTeam': ['Man City', 'West Ham', 'Middlesbrough', 'Southampton', 'Everton'],
    'FTHG': [4, 4, 1, 2, 2],
    'FTAG': [0, 2, 3, 2, 0],
    'HTHG': [2.0, 1.0, 1.0, 1.0, 2.0],
    'HTAG': [0.0, 0.0, 1.0, 2.0, 0.0],
    'HTR': ['H', 'H', 'D', 'A', 'H'],
    'HS': [17.0, 17.0, 6.0, 6.0, 17.0],
    'AS': [8.0, 12.0, 16.0, 13.0, 12.0],
    'HST': [14.0, 10.0, 3.0, 4.0, 8.0],
    'AST': [4.0, 5.0, 9.0, 6.0, 6.0],
    'HC': [6.0, 7.0, 8.0, 5.0, 6.0],
    'AC': [6.0, 7.0, 4.0, 8.0, 4.0],
    'HF': [13.0, 19.0, 15.0, 11.0, 21.0],
    'AF': [12.0, 14.0, 21.0, 13.0, 20.0],
    'HY': [1.0, 1.0, 5.0, 1.0, 1.0],
    'AY': [2.0, 2.0, 3.0, 1.0, 3.0],
    'HR': [0.0, 0.0, 1.0, 0.0, 0.0],
    'AR': [0.0, 0.0, 0.0, 0.0, 0.0]
}

# Create DataFrame
data = pd.DataFrame(data)


In [36]:
# Randomize the DataFrame
df_randomized = data.sample(frac=1, random_state=41).reset_index(drop=True)

# Display the randomized DataFrame
display(df_randomized)


HomeTeam       AwayTeam  FTHG  FTAG  HTHG  HTAG HTR    HS    AS   HST  AST  \
0   Chelsea       West Ham     4     2   1.0   0.0   H  17.0  12.0  10.0  5.0   
1  Coventry  Middlesbrough     1     3   1.0   1.0   D   6.0  16.0   3.0  9.0   
2     Leeds        Everton     2     0   2.0   0.0   H  17.0  12.0   8.0  6.0   
3     Derby    Southampton     2     2   1.0   2.0   A   6.0  13.0   4.0  6.0   
4  Charlton       Man City     4     0   2.0   0.0   H  17.0   8.0  14.0  4.0   

    HC   AC    HF    AF   HY   AY   HR   AR  
0  7.0  7.0  19.0  14.0  1.0  2.0  0.0  0.0  
1  8.0  4.0  15.0  21.0  5.0  3.0  1.0  0.0  
2  6.0  4.0  21.0  20.0  1.0  3.0  0.0  0.0  
3  5.0  8.0  11.0  13.0  1.0  1.0  0.0  0.0  
4  6.0  6.0  13.0  12.0  1.0  2.0  0.0  0.0

# Model Loading

In [37]:
# load model
with open('model.pkl', 'rb') as file:
    model = pickle.load(file)

# Prediction

### Prediction Match Result

In [38]:
# predict
predictions = model.predict(df_randomized)
print(predictions)

[0 1 0 2 0]


**Model Memprediksi:**
- Prediksi Match pertama adalah Home Win
- Prediksi match kedua adalah Away Win
- Prediksi Match ketiga adalah Home win
- Prediksi match keempat adalah Draw
- Prediksi match kelimat adalah Home win

### Prediction Odds for Match 

menggunakan predict_proba untuk mengeluarkan probabilitas terpilihnya kelas tersebut untuk menjadikan nilai probabilitas sebagai nilai dasar odds yang akan digunakan oleh agen taruhan atau betting agencies.

In [39]:
# predict
pred = model.predict_proba(df_randomized)
print(pred)

[[1.00 0.00 0.00]
 [0.00 1.00 0.00]
 [1.00 0.00 0.00]
 [0.00 0.00 1.00]
 [1.00 0.00 0.00]]


In [42]:
# Convert probabilities to betting odds
# Avoid division by zero 
epsilon = 1e-10
betting_odds = 1 / (pred + epsilon)

# Cap the maximum odds 
max_odds = 5  
betting_odds = np.clip(betting_odds, 0, max_odds)

# better readability
np.set_printoptions(suppress=True, formatter={'float_kind': '{:0.2f}'.format})

# Print the pred and the calculated betting odds
print("pred:\n", pred)
print("Betting Odds:\n", betting_odds)

# Set the random seed for reproducibility
np.random.seed(42)

# Adding decimal variation
odds_normalized = betting_odds + np.random.uniform(-0.50, 0.50, betting_odds.shape)
odds_normalized = np.round(odds_normalized, 2)

print("Betting Odds Normalized:")
odds_normalized1 = pd.DataFrame(odds_normalized)
display(odds_normalized1)

pred:
 [[1.00 0.00 0.00]
 [0.00 1.00 0.00]
 [1.00 0.00 0.00]
 [0.00 0.00 1.00]
 [1.00 0.00 0.00]]
Betting Odds:
 [[1.00 5.00 5.00]
 [5.00 1.00 5.00]
 [1.00 5.00 5.00]
 [5.00 5.00 1.00]
 [1.00 5.00 5.00]]
Betting Odds Normalized:


0     1     2
0  0.87  5.45  5.23
1  5.10  0.66  4.66
2  0.56  5.37  5.10
3  5.21  4.52  1.47
4  1.33  4.71  4.68

Maka odds dasar yang akan digunakan untuk betting adalah:
- Match 1
  - Home Win = 0.87
  - Away Win = 5.45
  - Draw = 5.23
- Match 2
  - Home Win = 5.10
  - Away Win = 0.66
  - Draw = 4.66
- Match 3
  - Home Win = 0.56
  - Away Win = 5.37
  - Draw = 5.10
- Match 4
  - Home Win = 5.21
  - Away Win = 4.52
  - Draw = 1.47
- Match 5
  - Home Win = 1.33
  - Away Win = 4.71
  - Draw = 4.68

Odds ini merupakan odds basic atau batas default sebuah odds sebelum diubah, maka betting agencies diperlukan untuk mengatur margin odds agar menghasilkan profitabilitas.